File: ogl-lab-02/house.cpp
#include <GLFW/glfw3.h>
#include <cmath>


// Draw the trunk of the tree using GL_QUADS
void drawTrunk()
{
    glBegin(GL_QUADS);
    glColor3f(0.6f, 0.3f, 0.0f); // Brown color for the trunk
    glVertex2f(-0.05f, -0.5f);
    glVertex2f(-0.05f, 0.0f);
    glVertex2f(0.05f, 0.0f);
    glVertex2f(0.05f, -0.5f);
    glEnd();
}

// Draw the leaves of the tree using GL_TRIANGLE_FAN
void drawLeaves()
{
    const int numTriangles = 12;
    const float radius = 0.2f;
    const float centerX = 0.0f;
    const float centerY = 0.0f;

    glBegin(GL_TRIANGLE_FAN);
    glColor3f(0.0f, 0.4f, 0.0f);  // Green color for the leaves
    glVertex2f(centerX, centerY); // Center of the leaves
    for (int i = 0; i <= numTriangles; ++i)
    {
        float angle = 2.0f * M_PI * static_cast<float>(i) / static_cast<float>(numTriangles);
        float x = centerX + radius * cos(angle);
        float y = centerY + radius * sin(angle);
        glVertex2f(x, y);
    }
    glEnd();
}

// Draw a tree using GL_TRIANGLES
void drawTree(float x, float y)
{
    glPushMatrix();
    glTranslatef(x, y, 0.0f); // Translate to the position of the tree
    drawTrunk();
    drawLeaves();
    glPopMatrix();
}

void drawSun()
{
    const int numTriangles = 36;
    const float radius = 0.1f;
    const float centerX = 0.5f;
    const float centerY = 0.8f;

    glBegin(GL_TRIANGLE_FAN);
    glColor3f(1.0f, 1.0f, 0.0f);  // Yellow color for the sun
    glVertex2f(centerX, centerY); // Center of the sun
    for (int i = 0; i <= numTriangles; ++i)
    {
        float angle = 2.0f * M_PI * static_cast<float>(i) / static_cast<float>(numTriangles);
        float x = centerX + radius * cos(angle);
        float y = centerY + radius * sin(angle);
        glVertex2f(x, y);
    }
    glEnd();
}

// Draw the landscape using GL_POLYGON
void drawLandscape()
{
    glBegin(GL_POLYGON);
    glColor3f(0.0f, 0.6f, 0.0f); // Green color for grass
    glVertex2f(-1.0f, -0.4f);
    glVertex2f(-1.0f, -1.0f);
    glVertex2f(1.0f, -1.0f);
    glVertex2f(1.0f, -0.4f);
    glEnd();
}

void drawHouse()
{
    // Draw the walls
    glBegin(GL_QUADS);
    glColor3f(0.8f, 0.7f, 0.6f); // Light brown color for walls
    glVertex2f(-0.3f, -0.4f);
    glVertex2f(-0.3f, 0.2f);
    glVertex2f(0.3f, 0.2f);
    glVertex2f(0.3f, -0.4f);
    glEnd();

    // Draw the roof
    glBegin(GL_TRIANGLES);
    glColor3f(0.9f, 0.1f, 0.1f); // Dark red color for roof
    glVertex2f(-0.35f, 0.2f);
    glVertex2f(0.0f, 0.5f);
    glVertex2f(0.35f, 0.2f);
    glEnd();

    // Draw the door
    glBegin(GL_QUADS);
    glColor3f(0.4f, 0.2f, 0.0f); // Brown color for door
    glVertex2f(-0.1f, -0.4f);
    glVertex2f(-0.1f, -0.1f);
    glVertex2f(0.1f, -0.1f);
    glVertex2f(0.1f, -0.4f);
    glEnd();

    // Draw the windows
    glBegin(GL_QUADS);
    glColor3f(0.9f, 0.9f, 1.0f); // Light blue color for windows
    glVertex2f(-0.2f, 0.0f);
    glVertex2f(-0.2f, 0.1f);
    glVertex2f(-0.05f, 0.1f);
    glVertex2f(-0.05f, 0.0f);

    glVertex2f(0.05f, 0.0f);
    glVertex2f(0.05f, 0.1f);
    glVertex2f(0.2f, 0.1f);
    glVertex2f(0.2f, 0.0f);
    glEnd();

    // Draw the chimney
    glBegin(GL_QUADS);
    glColor3f(0.4f, 0.2f, 0.1f); // Dark brown color for chimney
    glVertex2f(0.15f, 0.25f);
    glVertex2f(0.15f, 0.4f);
    glVertex2f(0.25f, 0.4f);
    glVertex2f(0.25f, 0.25f);
    glEnd();
}

void myDisplay(void)
{
    glClearColor(0.6f, 0.85f, 1.0f, 1.0f); // Light sky blue background
    glClear(GL_COLOR_BUFFER_BIT);

    drawHouse();
    drawSun();
    drawLandscape();
    drawTree(-0.8f, -0.2f);
    // drawTree(0.0f, -0.1f);
    drawTree(0.8f, -0.3f);
    // drawGrid();
    // drawAxes();

    glFlush();
}

int main()
{
    if (!glfwInit())
        return -1;

    GLFWwindow *window = glfwCreateWindow(800, 600, "Draw a House with scene", NULL, NULL);
    if (!window)
    {
        glfwTerminate();
        return -1;
    }

    glfwMakeContextCurrent(window);

    while (!glfwWindowShouldClose(window))
    {
        myDisplay();
        glfwSwapBuffers(window);
        glfwPollEvents();
    }

    glfwTerminate();
    return 0;
}

File: ogl-lab-02/house.py
import glfw
from OpenGL.GL import *
from math import pi, cos, sin


class HouseScene:
    def __init__(self, width: int, height: int, title: str) -> None:
        if not glfw.init():
            raise Exception("glfw can not be initialized!")

        self._win = glfw.create_window(width, height, title, None, None)

        if not self._win:
            glfw.terminate()
            raise Exception("glfw window can not be created!")

        glfw.set_window_pos(self._win, 400, 200)
        glfw.make_context_current(self._win)

    def draw_trunk(self):
        glBegin(GL_QUADS)
        glColor3f(0.6, 0.3, 0.0)  # Brown color for the trunk
        glVertex2f(-0.05, -0.5)
        glVertex2f(-0.05, 0.0)
        glVertex2f(0.05, 0.0)
        glVertex2f(0.05, -0.5)
        glEnd()

    def draw_leaves(self):
        num_triangles = 12
        radius = 0.2
        center_x, center_y = 0.0, 0.0

        glBegin(GL_TRIANGLE_FAN)
        glColor3f(0.0, 0.4, 0.0)  # Green color for the leaves
        glVertex2f(center_x, center_y)  # Center of the leaves
        for i in range(num_triangles + 1):
            angle = 2.0 * pi * float(i) / float(num_triangles)
            x = center_x + radius * cos(angle)
            y = center_y + radius * sin(angle)
            glVertex2f(x, y)
        glEnd()

    def draw_tree(self, x, y):
        glPushMatrix()
        glTranslatef(x, y, 0.0)  # Translate to the position of the tree
        self.draw_trunk()
        self.draw_leaves()
        glPopMatrix()

    def draw_sun(self):
        num_triangles = 36
        radius = 0.1
        center_x, center_y = 0.5, 0.8

        glBegin(GL_TRIANGLE_FAN)
        glColor3f(1.0, 1.0, 0.0)  # Yellow color for the sun
        glVertex2f(center_x, center_y)  # Center of the sun
        for i in range(num_triangles + 1):
            angle = 2.0 * pi * float(i) / float(num_triangles)
            x = center_x + radius * cos(angle)
            y = center_y + radius * sin(angle)
            glVertex2f(x, y)
        glEnd()

    def draw_landscape(self):
        glBegin(GL_POLYGON)
        glColor3f(0.0, 0.6, 0.0)  # Green color for grass
        glVertex2f(-1.0, -0.4)
        glVertex2f(-1.0, -1.0)
        glVertex2f(1.0, -1.0)
        glVertex2f(1.0, -0.4)
        glEnd()

    def draw_house(self):
        glBegin(GL_QUADS)
        glColor3f(0.8, 0.7, 0.6)  # Light brown color for walls
        glVertex2f(-0.3, -0.4)
        glVertex2f(-0.3, 0.2)
        glVertex2f(0.3, 0.2)
        glVertex2f(0.3, -0.4)
        glEnd()

        glBegin(GL_TRIANGLES)
        glColor3f(0.9, 0.1, 0.1)  # Dark red color for roof
        glVertex2f(-0.35, 0.2)
        glVertex2f(0.0, 0.5)
        glVertex2f(0.35, 0.2)
        glEnd()

        glBegin(GL_QUADS)
        glColor3f(0.4, 0.2, 0.0)  # Brown color for door
        glVertex2f(-0.1, -0.4)
        glVertex2f(-0.1, -0.1)
        glVertex2f(0.1, -0.1)
        glVertex2f(0.1, -0.4)
        glEnd()

        glBegin(GL_QUADS)
        glColor3f(0.9, 0.9, 1.0)  # Light blue color for windows
        glVertex2f(-0.2, 0.0)
        glVertex2f(-0.2, 0.1)
        glVertex2f(-0.05, 0.1)
        glVertex2f(-0.05, 0.0)

        glVertex2f(0.05, 0.0)
        glVertex2f(0.05, 0.1)
        glVertex2f(0.2, 0.1)
        glVertex2f(0.2, 0.0)
        glEnd()

        glBegin(GL_QUADS)
        glColor3f(0.4, 0.2, 0.1)  # Dark brown color for chimney
        glVertex2f(0.15, 0.25)
        glVertex2f(0.15, 0.4)
        glVertex2f(0.25, 0.4)
        glVertex2f(0.25, 0.25)
        glEnd()

    def my_display(self):
        glClearColor(0.6, 0.85, 1.0, 1.0)  # Light sky blue background
        glClear(GL_COLOR_BUFFER_BIT)

        self.draw_house()
        self.draw_sun()
        self.draw_landscape()
        self.draw_tree(-0.8, -0.2)
        self.draw_tree(0.8, -0.3)

        glFlush()

    def main_loop(self):
        while not glfw.window_should_close(self._win):
            self.my_display()
            glfw.swap_buffers(self._win)
            glfw.poll_events()

        glfw.terminate()


if __name__ == "__main__":
    win = HouseScene(800, 600, "Draw a House with Scene")
    win.main_loop()
